Add Microservice Transaction Sample (#5)
scalar-labs/scalardb-samples: microservice-transaction-sample/README.md

# Microservice Transaction Sample

> This is a sample application for Microservice Transaction that uses Two-phase Commit transactions in Scalar DB.

Reviewer: ```suggestion
This is a sample application for Microservice Transaction that uses Two-phase Commit Transactions in Scalar DB.
```

> You can find more information about Two-phase Commit transactions in Scalar DB [here](https://github.com/scalar-labs/scalardb/tree/master/docs/two-phase-commit-transactions.md).

Reviewer: ```suggestion
You can find more information about Two-phase Commit Transactions in Scalar DB [here](https://github.com/scalar-labs/scalardb/tree/master/docs/two-phase-commit-transactions.md).
```

## Prerequisites
- Java (OpenJDK 8 or higher)
- Gradle
- Docker, Docker Compose

## Sample application

### Overview

> There are two microservices called *Customer Service* and *Order Service* in this sample application.

Reviewer: ```suggestion
There are two microservices called *Customer Service* and *Order Service* based on the [*Database-per-service* pattern](https://microservices.io/patterns/data/database-per-service.html) in this sample application.
```

> Customer Service manages the customer's information, including credit card information like a credit limit and a credit total.

Reviewer: ```suggestion
Customer Service manages customers' information including credit card information like a credit limit and a credit total.
```

Order Service is responsible for order operations like placing an order and getting order histories.

> Each service has gRPC endpoints, and clients and the services call them.

Reviewer: ```suggestion
Each service has gRPC endpoints. Clients call the endpoints, and the services call the endpoints each other as well.
```

> Customer Service and Order Service uses MySQL and Cassandra, respectively.

Reviewer: ```suggestion
Customer Service and Order Service uses MySQL and Cassandra through Scalar DB, respectively.
```

![Overview](images/overview.png)

> Also, both services commonly access the Coordinator database that's used in the Consensus Commit protocol.

Reviewer: ```suggestion
Note that both services access a small coordinator database used for the Consensus Commit protocol.
The database is service-independent and exists for managing transaction metadata for Consensus Commit in a highly available manner, so we don't think it is violating the database-per-service pattern. We also plan to create a microservice container for the coordinator database to truly achieve the database-per-service pattern.
```

Reviewer: The second sentence can/should be in a *NOTE* or something.

> In this sample application, the Coordinator database uses Cassandra, also used in Order Service, but you can also use a dedicated database for it in your application.

Reviewer: ```suggestion

In this sample application, for ease of setup and explanation, we colocate the coordinator database in the same Cassandra instance of the Order Service, but of course, the coordinator database can be managed as a separate database.
```

> All data accesses in the microservices are done through Scalar DB.

Reviewer: ```suggestion
```
Added Scalar DB description above.

### Schemas

[The schema for Customer Service](customer-service-schema-loader/customer-service-schema.json) is as follows:

```json
{
  "customer_service.customers": {
    "transaction": true,
    "partition-key": [
      "id"
    ],
    "clustering-key": [],
    "columns": {
      "id": "INT",
      "name": "TEXT",
      "credit_limit": "INT",
      "credit_total": "INT"
    }
  }
}
```

> The `customer_service.customers` table manages the customer information.

Reviewer: ```suggestion
The `customer_service.customers` table manages customers' information.
```

> The `credit_limit` column is the maximum amount of money a lender will allow a customer to spend using a credit card.

Reviewer: ```suggestion
The `credit_limit` for a customer is the maximum amount of money a lender will allow the customer to spend using a credit card,
```

> And, the `credit_total` column is the amount of money that a customer has already spent by using a credit card.

Reviewer: ```suggestion
and the `credit_total` is the amount of money that the customer has already spent by using the credit card.
```

[The schema for Order Service](order-service-schema-loader/order-service-schema.json) is as follows:

```json
{
  "order_service.orders": {
    "transaction": true,
    "partition-key": [
      "customer_id"
    ],
    "clustering-key": [
      "timestamp"
    ],
    "secondary-index": [
      "id"
    ],
    "columns": {
      "id": "TEXT",
      "customer_id": "INT",
      "timestamp": "BIGINT"
    }
  },
  "order_service.statements": {
    "transaction": true,
    "partition-key": [
      "order_id"
    ],
    "clustering-key": [
      "item_id"
    ],
    "columns": {
      "order_id": "TEXT",
      "item_id": "INT",
      "count": "INT"
    }
  },
  "order_service.items": {
    "transaction": true,
    "partition-key": [
      "id"
    ],
    "clustering-key": [],
    "columns": {
      "id": "INT",
      "name": "TEXT",
      "price": "INT"
    }
  }
}
```

> The `order_service.orders` table manages the order data, and the `order_service.statements` table contains the statement data of the order.

Reviewer: ```suggestion
The `order_service.orders` table manages orders' information, and the `order_service.statements` table manages the states' information of the orders.
```
BTW, statement is the correct wording for the context? states or statuses?

> The `order_service.items` table maintains the item data.

Reviewer: ```suggestion
The `order_service.items` table manages items' information to be ordered.
```

  }
}
```

The `customer_service.customers` table manages customers' information.
The `credit_limit` for a customer is the maximum amount of money a lender will allow the customer to spend using a credit card,
and the `credit_total` is the amount of money that the customer has already spent by using the credit card.

[The schema for Order Service](order-service-schema-loader/order-service-schema.json) is as follows:

```json
{
  "order_service.orders": {
    "transaction": true,
    "partition-key": [
      "customer_id"
    ],
    "clustering-key": [
      "timestamp"
    ],
    "secondary-index": [
      "id"
    ],
    "columns": {
      "id": "TEXT",
      "customer_id": "INT",
      "timestamp": "BIGINT"
    }
  },
  "order_service.statements": {
    "transaction": true,
    "partition-key": [
      "order_id"
    ],
    "clustering-key": [
      "item_id"
    ],
    "columns": {
      "order_id": "TEXT",
      "item_id": "INT",
      "count": "INT"
    }
  },
  "order_service.items": {
    "transaction": true,
    "partition-key": [
      "id"
    ],
    "clustering-key": [],
    "columns": {
      "id": "INT",
      "name": "TEXT",
      "price": "INT"
    }
  }
}
```

> The `order_service.orders` table manages orders' information, and the `order_service.statements` table manages the states' information of the orders.

Reviewer: `statements` table is OK?

Author: Oh sorry, I missed this question in your previous review.

I use `statements` as `明細` in Japanese. But I'm not sure this wording is correct honestly.

Reviewer: Oh, OK, then my previous suggestion is wrong.

        at example.rpc.OrderServiceGrpc$OrderServiceBlockingStub.placeOrder(OrderServiceGrpc.java:295)
        at example.client.command.PlaceOrderCommand.call(PlaceOrderCommand.java:35)
        at example.client.command.PlaceOrderCommand.call(PlaceOrderCommand.java:12)
        at picocli.CommandLine.executeUserObject(CommandLine.java:1783)
        at picocli.CommandLine.access$900(CommandLine.java:145)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2141)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2108)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:1975)
        at picocli.CommandLine.execute(CommandLine.java:1904)
        at example.client.Client.main(Client.java:27)
...
```

After repayment, the customer will be able to place an order again!

```
# ./gradlew :client:run --args="Repayment 1 8000"
...
# ./gradlew :client:run --args="GetCustomerInfo 1"
...
{"id": 1,"name": "Yamada Taro","credit_limit": 10000,"credit_total": 2000}
...
# ./gradlew :client:run --args="PlaceOrder 1 3:1,4:1"
...
{"order_id": "dd53dd9d-aaf4-41db-84b2-56951fed6425"}
...

> ```

Reviewer: I think we need to explain where and how the microservice transaction is implemented roughly.
I think it can be simple and the detail how part can be delegated to the two-phase commit guide.

Something like this.
```
## How the microservice transaction is achieved

So far, we have run the sample application but we haven't seen how it is implemented, especially for the transaction that spans the services.
Let's look at the code to see how it is implemented.
...
```

  }
}
```

The `customer_service.customers` table manages customers' information.
The `credit_limit` for a customer is the maximum amount of money a lender will allow the customer to spend using a credit card,
and the `credit_total` is the amount of money that the customer has already spent by using the credit card.

[The schema for Order Service](order-service-schema-loader/order-service-schema.json) is as follows:

```json
{
  "order_service.orders": {
    "transaction": true,
    "partition-key": [
      "customer_id"
    ],
    "clustering-key": [
      "timestamp"
    ],
    "secondary-index": [
      "id"
    ],
    "columns": {
      "id": "TEXT",
      "customer_id": "INT",
      "timestamp": "BIGINT"
    }
  },
  "order_service.statements": {
    "transaction": true,
    "partition-key": [
      "order_id"
    ],
    "clustering-key": [
      "item_id"
    ],
    "columns": {
      "order_id": "TEXT",
      "item_id": "INT",
      "count": "INT"
    }
  },
  "order_service.items": {
    "transaction": true,
    "partition-key": [
      "id"
    ],
    "clustering-key": [],
    "columns": {
      "id": "INT",
      "name": "TEXT",
      "price": "INT"
    }
  }
}
```

> The `order_service.orders` table manages orders' information, and the `order_service.statements` table manages the states' information of the orders.

Reviewer: ```suggestion
The `order_service.orders` table manages orders' information, and the `order_service.statements` table manages the statements' information of the orders.
```
